Zend/zend_closures.c

```diff
@@ -381,7 +381,45 @@ static ZEND_COLD zend_function *zend_closure_get_constructor(zend_object *object
 static int zend_closure_compare(zval *o1, zval *o2) /* {{{ */
 {
 	ZEND_COMPARE_OBJECTS_FALLBACK(o1, o2);
-	return Z_OBJ_P(o1) != Z_OBJ_P(o2);
+
+	zend_closure *lhs = (zend_closure*) Z_OBJ_P(o1);
+	zend_closure *rhs = (zend_closure*) Z_OBJ_P(o2);
+
+	if (lhs == rhs) {
+	    return SUCCESS;
+	}
+
+	if (!((lhs->func.common.fn_flags & ZEND_ACC_FAKE_CLOSURE) && (rhs->func.common.fn_flags & ZEND_ACC_FAKE_CLOSURE))) {
+		return ZEND_UNCOMPARABLE;
+	}
+
+	if (Z_TYPE(lhs->this_ptr) == IS_OBJECT && Z_TYPE(rhs->this_ptr) == IS_OBJECT) {
```

Comment:
Do we need to check whether object-ness of lhs and rhs this_ptr matches? Apparently we currently allow writing this code: https://3v4l.org/odRMs (Though we probably shouldn't...)

Reply:
No, fixed in dc48bd905205c035d3fac2ad931660d2aca1fe91